File: ruby/lib/cekat_event_sdk/visitor_id_resolver.rb
# frozen_string_literal: true

module CekatEventSdk
  # Resolves the browser visitor ID: a nonblank X-Cekat-Visitor-ID header wins over a nonblank
  # _cekat_visitor_id cookie. Values are trimmed and otherwise not decoded.
  module VisitorIdResolver
    module_function

    # Trims a visitor ID; blank or non-String values become nil.
    def normalize(visitor_id)
      return nil unless visitor_id.is_a?(String)

      trimmed = visitor_id.strip
      trimmed.empty? ? nil : trimmed
    end

    def from_header_and_cookie(header:, cookie:)
      normalize(header) || normalize(cookie)
    end

    # Returns the first nonblank _cekat_visitor_id value in a raw Cookie header.
    def from_cookie_header(cookie_header)
      return nil unless cookie_header.is_a?(String)

      cookie_header.split(";").each do |pair|
        name, separator, value = pair.partition("=")
        next if separator.empty? || name.strip != VISITOR_COOKIE

        visitor_id = normalize(value)
        return visitor_id if visitor_id
      end
      nil
    end

    # Resolves from a Rack env. The raw Cookie header is used so values are never unescaped.
    def from_rack_env(env)
      from_header_and_cookie(header: env["HTTP_X_CEKAT_VISITOR_ID"], cookie: from_cookie_header(env["HTTP_COOKIE"]))
    end
  end
end

File: dotnet/src/Cekat.EventSdk/Visitor/VisitorIdResolver.cs
namespace Cekat.EventSdk;

/// <summary>Applies the visitor precedence rules shared by every Cekat SDK.</summary>
public static class VisitorIdResolver
{
    /// <summary>Returns the trimmed value, or <see langword="null"/> when blank.</summary>
    public static string? Normalize(string? visitorId)
    {
        var trimmed = visitorId?.Trim();
        return string.IsNullOrEmpty(trimmed) ? null : trimmed;
    }

    /// <summary>A nonblank header value wins over a nonblank cookie value.</summary>
    public static string? FromHeaderAndCookie(string? header, string? cookie) => Normalize(header) ?? Normalize(cookie);

    /// <summary>Returns the first nonblank header value, then the first nonblank cookie value.</summary>
    public static string? FromHeadersAndCookies(IEnumerable<string?> headers, IEnumerable<string?> cookies)
    {
        ArgumentNullException.ThrowIfNull(headers);
        ArgumentNullException.ThrowIfNull(cookies);
        return headers.Select(Normalize).FirstOrDefault(value => value is not null)
            ?? cookies.Select(Normalize).FirstOrDefault(value => value is not null);
    }

    /// <summary>
    /// Returns every <c>_cekat_visitor_id</c> value in a raw <c>Cookie</c> header, undecoded, in order.
    /// </summary>
    public static IEnumerable<string> CookieValues(string? cookieHeader)
    {
        if (string.IsNullOrEmpty(cookieHeader))
        {
            yield break;
        }

        foreach (var pair in cookieHeader.Split(';'))
        {
            var separator = pair.IndexOf('=', StringComparison.Ordinal);
            if (separator > 0 && pair.AsSpan(0, separator).Trim().SequenceEqual(CekatConstants.VisitorCookie))
            {
                yield return pair[(separator + 1)..];
            }
        }
    }

    /// <summary>A nonblank explicit visitor ID wins over the context.</summary>
    public static string? ForEvent(string? explicitVisitorId, IVisitorContext context)
    {
        ArgumentNullException.ThrowIfNull(context);
        return Normalize(explicitVisitorId) ?? Normalize(context.CurrentVisitorId);
    }
}
